### ISSUE TITLE
Possibility to have a strict dayfirst

### ISSUE THREAD

**Issue by jorisvandenbossche** [dateutil/dateutil#214]
pandas uses the `dateutil.parser.parse` functions in its [`to_datetime`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html) method (and it's what makes this a powerful method to be able to parse all kind of sorts datetime strings!)

But the pandas `to_datetime` function can parse lists (or arrays) of strings at once, and a regular bug report we get is that you can end up with inconsistent parsing within that list (because of the flexibility of dateutil). 
For example:

```
In [20]: pd.to_datetime(["31/12/2014", "10/03/2011"])
Out[20]: DatetimeIndex(['2014-12-31', '2011-10-03'])
```

So in the above, both strings are parsed in another way. Of course, in this case it can be solved by passing `dayfirst=True` (which we pass through to `dateutil.parser.parse`). But say you have a malformed string:

```
In [21]: pd.to_datetime(["31/12/2014", "10/13/2011"], dayfirst=True)
Out[21]: DatetimeIndex(['2014-12-31', '2011-10-13'])
```

Regardless of the `dayfirst=True` that is passed, not all strings are parsed with using day first, but you end up with a mixture of using dayfirst or not depending on the string.

So our **question is if it would be possible to be able to specify that dayfirst/yearfirst have to be strict** (for example with an extra keyword), so that if the provided string cannot be parsed following the given dayfirst/yearfirst argument, that an error is raised instead of parsing the string anyway without following dayfirst/yearfirst.

The original issue is here: https://github.com/pydata/pandas/issues/3341

**Comment by pganssle** [comment]
Hm. I suppose this is possible, but I'm a bit concerned about the proliferation of options passed directly to the parser like this, which is why I rolled back the `smart_defaults` and `fallback_on_invalid_dates` options in #191 before the 2.5.0 release.

In my opinion, a lot of the proliferation of these command line arguments could be mitigated by giving users access to the date data *before* the date is constructed. Ideally, this would even allow you to generate an `strptime` string on the first run in `pandas`, which can then be applied to the remainder of the dates (and you can re-run the parser if you get a `ValueError`.

@jorisvandenbossche Obviously it'll take a bit longer than implementing a "strict" dayfirst, but do you think something like this will be an acceptable solution?

**Comment by jorisvandenbossche** [comment]
@pganssle Thanks for your quick reply!

> I suppose this is possible, but I'm a bit concerned about the proliferation of options passed directly to the parser like this

That's something I certainly understand

> In my opinion, a lot of the proliferation of these command line arguments could be mitigated by giving users access to the date data before the date is constructed. Ideally, this would even allow you to generate an strptime string on the first run in pandas, which can then be applied to the remainder of the dates (and you can re-run the parser if you get a ValueError).

I am not sure if this would fix the issues for pandas (but besides that, something like that could certainly be useful).

To give a simplified overview of what pandas does in `to_datetime` for parsing strings to datetimes:

- if the user passed a `format` (like in `strptime`), then pandas uses its own (faster) parser
- if the user passed `infer_datetime_format=True` (default is now still False, but we want to change this), the format is inferred from the first value and if, using this inferred format, all values can be parsed without errors, this result is returned.
- try parsing the strings as ISO-like strings. If all values can be parsed without error, return this result
- If the above fails (this means: no format provided or inferred and not consistenly ISO-like format), then use the flexible `dateutil` parser

So we already have some functionality to infer the format and used that specification.

But we use `dateutil` if no format is defined, or if the format is not consistent. And it is actually a feature of pandas currently (relying on dateutil) that `to_datetime` *can* still handle messy data with inconsistent format eg ``['2012-01-01', '2012/01/02']``, which not be parsed by providing a `format`. 
So we actually want to use this flexibility, but at the same time be able to guarantee that the order of day/month *is* consistent / respects what is specified with `dayfirst`.

**Comment by pganssle** [comment]
@jorisvandenbossche Yes, thanks for the information - I actually do use `pandas` pretty heavily on a day-to-day basis, so I was familiar with a lot of this, but it's nice to have it clarified here.

> * if the user passed `infer_datetime_format=True` (default is now still False, but we want to change this), the format is inferred from the first value and if, using this inferred format, all values can be parsed without errors, this result is returned.

Yes, I was hoping that this would be some functionality we could add to dateutil, since the parser is *already* inferring the date format - so you'd basically get this for free just using the dateutil parser.

I will think about the exact details of the implementation and come back with some options. I don't think it'll be a huge deal to add some sort of strictness component to `parserinfo` objects.

**Comment by pganssle** [comment]
OK, I've thought this over and I'm thinking that we'll start by adding an `order_strict` flag, defaulting to `False` in the `parserinfo` object. If `yearfirst` or `dayfirst` is specified (i.e. not `None`), the flag will have no effect, otherwise, it will throw a `ValueError` if the `order_strict` flag is set to `False`.

I also think that the more sensible default is that `order_strict` be set to `True`, since if you're actually going through the trouble of specifying whether year or day comes first, then you probably don't want it to silently ignore that if the number it gets is wrong - please let me know if you disagree about this, because I don't actually use the parameter myself. If there's agreement that `True` is the sensible default, I'll add in a `FutureWarning` that the default will change, and the default will change in a future version (we already have a few `DeprecationWarning`s floating around, so I figure we'll accumulate them over the next few feature releases, then do a major compatibility-breaking 3.0 release in a year or so that actually acts on the threats).

**Comment by subodh-malgonde** [comment]
> I also think that the more sensible default is that order_strict be set to True

@pganssle I agree with you. 

This thread has been silent for a long time, any progress or conclusion about having a strict dayfirst?
